### PR TITLE
Scrollbar markers enhancements

### PR DESCRIPTION
**Pull request by alygin**
Several improvements in how various markers are displayed in the editor scrollbar, as described in #9070, if you're ok with the proposal:
- Scrollbar has three columns:
  - 1st is for git markers
  - 2nd is for selections and search resulta highlightings
  - 3rd is for diagnostics
- Height of all markers is scaled, but there's a minimal allowed height of 2px.
- Right border removed from both the scrollbar and thumb to make more room for markers.

Release Notes:

- Improved scrollbar markers visualization (#9070).

### CODE REVIEW

**Comment by evrsen** [comment]
Could this be an opt-in option instead?

**Comment by alygin** [comment]
> Could this be an opt-in option instead?

Do you mean this type of visualization, so you could switch back to the current approach? Or would you like to be able to switch scrollbar markers off completely?

If the former, let's discuss it in the linked issue (#9070). I'll be glad to hear, what's bad with this approach, maybe we'll find a better one.

If the latter, you already can do that via the `scrollbar` settings block. Here're the options: [default.json#135](https://github.com/zed-industries/zed/blob/456efb53ad6dc814012818a222e04b2e290fbeb0/assets/settings/default.json#L135-L141)

**Comment by alygin** [comment]
@evrsen, you're right, this PR intentionally doesn't introduce new markers, it focuses on improving already existing ones, so we have a clear consensus on what approach we're going to use for such things.

There may be lots of further enhancements, like markers for cursors, markers for todo-comments, enclosing braces markers separtaion from the current symbol highligtings (to reduce noise), etc. But first things first :)

**Review by SomeoneToIgnore — COMMENTED** [review]
I like what I see now much better, thank you for prototyping.
Screenshot 2024-03-11 at 12 03 42

I played with it a bit and found that our issues with the scroll bar had somehow worsened?

* Consider searching for `option` in the [project.rs](https://github.com/zed-industries/zed/blob/main/crates/project/src/project.rs) file, clicking the marks on the scrollbar and looking where the screen display on the scrollbar moves.
I've also spoiled the `LanguageServer` use to `Language@Server` to induce more error marks in the same file, but I think there are better ways to produce more marks.

But even with the current results, consider line # 550 in your current patch
image

Previous `option` match is on line 485, next one is on 979, the editor range on the screen does not contain any of the matches but has some false-positive errors from r-a.

In the screen range painted on the scrollbar, I would expect to have no search result marks and two error ones.
Both scrolls fail my expectations:
image

but the original one makes more sense to me? There are definitely lines (> 0) between `option` query matches, and the last match is rather away from the error marks on the screen.

* optional: seems that it's broken already, but to mention — clicking on the project.rs scrollbar methodically from top to bottom or reverse seems to move the screen very abruptly, sometimes the moved screen range is not even touching the point clicked.

I think, there's some coordinate issue at play here, most probably buffer and display rows mixed. I do not think it's relevant to this PR, but wanted to mention since the previous bullet is somewhat related.

**Comment by alygin** [comment]
@SomeoneToIgnore, 

> I would expect to have no search result marks and two error ones.

In such a large file, the real scroll thumb height is one pixel, so Zed imposes a minimal height: https://github.com/zed-industries/zed/blob/2b67bb27cf87650b25f7ae25f43539dba10f3b77/crates/editor/src/element.rs#L946-L953

But unlike other editors, it tries to synchronize the center of the thumb with the center of the editor. So, you get lots of markers both up and down from the thumb center, for the lines that are really invisible in the editor. Another confusing case is when you have a marker for the first row in a large file. It sits in the middle of the thumb scrolled all the way up.

Other editors match the top border of the thumb with the top visible row. It doesn't prevent false marker appearence iside the thumb, but it makes such problems less vivid. I would suggest to use the same approach.

> clicking on the project.rs scrollbar methodically from top to bottom or reverse seems to move the screen very abruptly, sometimes the moved screen range is not even touching the point clicked

It looks like the mouse click handler doesn't take into account all those adjustments made for very small thumbs.

None of these is introduced by the PR. I think I can fix it all. But should it be done in the same PR?

**Comment by alygin** [comment]
> In such a large file, the real scroll thumb height is one pixel, so Zed imposes a minimal height:

BTW, it also has this effect of this thumb, moving out of the scrollbar:

https://github.com/zed-industries/zed/assets/2101250/9d8511e7-e339-4689-a1d1-a66eb71b69c8

**Comment by SomeoneToIgnore** [comment]
Oh, good to know that the first one is also not something new.
In this case, looks like this version can be merged in, unless @maxdeviant has some ideas?

I would not merge this pr before Wednesday, as there's a bunch of big changes going on already this week, and we might want to with for another, visual change before the next Preview, just in case.

As for the two issues mentioned, I do not think this PR needs to address any of these, but would be very nice to get into some day later.

**Review by SomeoneToIgnore — APPROVED** [review]
Let's get this in and get more feedback on the next Nightly, thank you for making an effort to improve the visuals there.

If this new approach is not reverted, we might consider fixing those exaggerated scrollbar issues discussed.

**Comment by JosephTLyons** [comment]
Hey @alygin, I dont know what all sort of technical challenges you ran into while doing this, and so maybe some of this comes down to a choice you made, but I noticed that the error markers are a bit longer than the others and overlap a tiny bit with the right border of the scrollbar:

SCR-20240314-mbgt

**Comment by alygin** [comment]
@JosephTLyons,

An eagle eye :)

This is the effect of the following facts and decisions:
1. The scrollbar in Zed is 12px wide including both borders.
2. The first pixel is reserved for the left scrollbar border. Unlike the right border, it shouldn't be used to draw markers because when the thumb overlaps it, this left pixel becomes invisible, which leads to a visual effect I wanted to avoid:

zed-scroll-left

3. So, we have 11px to fit all the markers into.
4. I decided to give all three marker columns equal width, but if the full width is not divisable by 3, the reminder goes to the diagnostics column because this is usually the most important thing we want to see (in my opinion :).

Thus, given 11px for the full width, we end up with 3px + 3px + 5px.

Do you see a room for improvements here? Those things are easy to change and compare results.

**Comment by alygin** [comment]
If we don't draw the thumb's left border, we can remove that negative left-border effect, and have same-width markers:

scroll-12px

**Comment by alygin** [comment]
> Let's get this in and get more feedback on the next Nightly

@SomeoneToIgnore, @JosephTLyons, how is it going so far? What do you think of the change from my previous comment? I could raise a PR with this fix.

**Comment by SomeoneToIgnore** [comment]
I like how it looks, no objections from my side to move on.
I think the comment above with the fix is a good thing to do, also the click & scroll broken part is something I would love to get fixed.

@ConradIrwin may have some opinion to share on the design too.

**Comment by alygin** [comment]
> also the click & scroll broken part is something I would love to get fixed.

The click & scroll fix will depend on the decision about the thumb positioning approach &mdash; will we keep the current "thumb center to screen center" behaviour, or switch to the "thumb top to screen top" which is used by other editors?

**Comment by alygin** [comment]
> Also, there seems to be a perf issue with ~5k search entries displayed on the scrollbar, @mrnugget was looking into that, but welcome to join the effort

Should be easy to fix. Give me a second.

**Comment by alygin** [comment]
@SomeoneToIgnore, @mrnugget, something like this should do the job: https://github.com/alygin/zed/pull/4/files

It's just a quick-and-dirty implementation. I can make it more elegant if it works in your 5k-search-entries case.

**Comment by mrnugget** [comment]
Hey! Full context is in this issue here: https://github.com/zed-industries/zed/issues/9055#issuecomment-1999318878 You can see in the error messages one user posted that we're rendering too many quads. The idea was that we can quantize those probably and render less (because who can see 50k results on a scrollbar?).

Your PR makes sense to me. I'm wondering what it's impact is on a practical level, i.e. how many quads would we draw less?

**Comment by alygin** [comment]
> how many quads would we draw less?

This is highly depends on the density of the markers. The idea is to display a single quad instead of a group when they overlap. For instance, if i search the letter "s" in Zed's `project.rs`, I get ~11K matches but only draw 7 quads.

We can also increase density by increasing the `MIN_MARKER_HEIGHT`. It will lead to even less quads.

**Comment by mrnugget** [comment]
That seems like a good improvement then. Would be nice to test it with a very long file so we get more lines in the scrollbar. I can do that if you want.

**Comment by alygin** [comment]
If you have some reference results we can compare against, it would be interesting to see if there’s effect from such optimization.

Though [we discussed this part](https://github.com/alygin/zed/pull/4#pullrequestreview-1943780146) with @osiewicz yesterday, and he wisely pointed out that there’s another heavy lifting that we do on every scrollbar redrawing. I would expect the main performance boost from eliminating it by caching. I’ll try to do this later today.

**Comment by mrnugget** [comment]
1. Download https://drive.google.com/file/d/1Y7qh7wsPkb0dhVRj9L5-J5CruQXmpt36/view?usp=sharing (it's a log from our GitHub action run concatenated a bunch of times to make it ~150mb)
2. Open in Zed
3. Make sure scrollbar is on
4. Search for `T08`

On `main`, I get the following error in console:

```
[2024-03-19T11:42:23+01:00 ERROR gpui::platform::mac::metal_renderer] scene too large: 85 paths, 0 shadows, 68305 quads, 0 underlines, 4960 mono, 0 poly, 0 surfaces
```

And rendering essentially breaks down.

**Comment by alygin** [comment]
It doesn't fail on my brach. Though it doesn't work fast either.
We surely need more improvements than just combining quads.

**Comment by mrnugget** [comment]
Did you try it in `--release` mode? That makes a huge difference with this stuff.

**Comment by alygin** [comment]
Yes, the scrolling is slow even in the release mode.

**Comment by alygin** [comment]
It looks like we're painting search results twice because `background_highlights_in_range` [here](https://github.com/zed-industries/zed/blob/269d2513ca3c7d4aa8a2ce17e71841eaa6dc2b01/crates/editor/src/element.rs#L2220) returns all the ranges regardles of their types.

I've removed the whole "selections" block in my branch and it still marks all the selections. I think we need to fix that.

I've also [added a method](https://github.com/alygin/zed/blob/d1a7019c251fa4f4769e1fa0bcbd5b3a12dba034/crates/editor/src/editor.rs#L9232) that returns ranges of row numbers to highlight, just to experiment. It's very close to `background_highlight_row_ranges()`, yet a bit more optimized for the scrollbar. It's somewhat faster, and I can work with that large file with 68K search results on my branch, though the scrolling is not smooth yet.

The next stop is caching.

**Comment by mrnugget** [comment]
@alygin do you maybe want to open a draft PR and we can continue the discussion there? I think that makes it easier to discuss things — next to the code.

**Comment by alygin** [comment]
@mrnugget, sure, [here's the PR in my repo](https://github.com/alygin/zed/pull/4) where we already started the discussion. I didn't want to clutter Zed's repo with such "playground PR's", but I can move it here if it's OK.

**Comment by alygin** [comment]
> Yeah, I think that's OK. Makes it easier to check out the PR, IMHO.

Sure, here it is: #9647

**Pull request merged**
